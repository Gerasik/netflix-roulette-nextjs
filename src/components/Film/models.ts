import { MovieMap } from 'models';
import * as FilmModels from 'containers/Film/models';

export interface FilmProps {
  filmData: MovieMap;
  setFilmId: FilmModels.ActionCreator.SetFilmId;
}
