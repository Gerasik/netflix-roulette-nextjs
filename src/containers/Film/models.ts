import * as Models from 'models';

export type State = Models.MovieMap;

export namespace Payload {
  export type SetFilmData = Models.MovieMap;
  export type SetFilmId = number;
}

export namespace Action {
  export type SetFilmData = Models.IAction<Payload.SetFilmData>;
  export type SetFilmId = Models.IAction<Payload.SetFilmId>;
}

export namespace ActionCreator {
  export type SetFilmData = (payload: Payload.SetFilmData) => Action.SetFilmData;
  export type SetFilmId = (payload: Payload.SetFilmId) => Action.SetFilmId;
}
