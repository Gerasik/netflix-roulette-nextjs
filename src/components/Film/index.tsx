import React, { useEffect, FunctionComponent, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useRouter } from 'next/router';

import imagePlaceholder from 'assets/image_placeholder.png';
import { FilmProps } from './models';

import styles from './style.module.scss';
import { fromJS } from 'immutable';

const Film = ({ filmData, setFilmId }) => {
  const router = useRouter();
  const { id } = router.query;

  useEffect(() => {
    setFilmId(Number(id));
    window.scrollTo(0, 0);
  }, [id]);

  const [imgError, setImgError] = useState(true);

  if (!filmData || !filmData.has('id')) return <span> No data found</span>;

  const imagePath = filmData.get('poster_path');
  const poster_path = imagePath || imagePlaceholder;
  const title = filmData.get('title');
  const rating = filmData.get('vote_average');
  const smallDescription = filmData.get('tagline');
  const year = new Date(filmData.get('release_date')).getFullYear();
  const runtime = filmData.get('runtime') || 0;
  const overview = filmData.get('overview');

  const image = imgError ? (
    <picture className={styles.picture} style={imagePath ? { backgroundColor: 'inherit' } : {}}>
      <source srcSet={poster_path} media="(min-width: 600px)" />
      <img
        src={poster_path}
        width={imagePath ? '100%' : '50%'}
        alt={title}
        onError={(): void => setImgError(true)}
      />
    </picture>
  ) : (
    <picture className={styles.picture}>
      <source srcSet={imagePlaceholder} media="(min-width: 600px)" />
      <img src={imagePlaceholder} width="50%" alt="error load" />
    </picture>
  );

  return (
    <div className={styles.container}>
      {image}
      <div className={styles.info}>
        <div className={styles['title-container']}>
          <h1 className={styles.title}>{title}</h1>
          <div className={styles.round}>
            <span className={styles.rating}>{rating}</span>
          </div>
        </div>
        <span className={styles.smallDescription}>{smallDescription}</span>
        <div className={styles.parameters}>
          <span className={styles.year}>{year}</span>
          <span className={styles.runtime}>{`${runtime} min`}</span>
        </div>
        <p className={styles.overview}>{overview}</p>
      </div>
    </div>
  );
};

export default Film;
