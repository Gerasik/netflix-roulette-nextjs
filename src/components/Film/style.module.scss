@import 'styles/colors.scss';

.container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  color: white;
  .picture {
    min-width: 15em;
    min-height: 23em;
    width: 20%;
    background-color: rgba($color: $fontColor, $alpha: 0.7);
    display: flex;
    justify-content: center;
    align-items: center;
  }
  .info {
    opacity: 0.8;
    width: 75%;
    font-size: 1em;
    .title-container {
      display: flex;
      align-items: center;
      font-size: 1.4em;
      margin: 0.5em 0 0.3em;

      .title {
        display: inline-block;
        font-weight: normal;
        letter-spacing: 0.1rem;
        margin: 0 0.5em 0.1em 0;
      }
      .round {
        border: 0.1rem solid $fontColor;
        width: 3rem;
        height: 3rem;
        border-radius: 50%;
        display: flex;
        justify-content: center;
        align-items: center;
        font-size: 1rem;
      }
      .rating {
        font-size: 1.3rem;
        color: #a1e66f;
        display: block;
      }
    }
    .smallDescription {
      margin: 0.2em 0 2em;
      display: inline-block;
    }
    .parameters {
      color: $secondaryFontColor;
      font-size: 1.5rem;
      font-weight: 500;
      .year {
        margin-right: 0.5em;
      }
    }
    .overview {
      font-size: 1rem;
    }
  }
}
