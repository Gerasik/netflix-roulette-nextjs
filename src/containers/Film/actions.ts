import ActionType from './constants';
import { ActionCreator } from './models';

export const setFilmData: ActionCreator.SetFilmData = payload => ({
  type: ActionType.SET_FILM_DATA,
  payload,
});

export const setFilmId: ActionCreator.SetFilmId = payload => ({
  type: ActionType.SET_FILM_ID,
  payload,
});
