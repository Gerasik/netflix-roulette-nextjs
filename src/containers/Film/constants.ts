const ActionType = {
  SET_FILM_DATA: 'SET-FILM-DATA',
  SET_FILM_ID: 'SET-FILM-ID',
};

export default ActionType;
