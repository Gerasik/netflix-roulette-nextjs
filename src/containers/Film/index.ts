import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';

import Film from 'components/Film';
import { setFilmId } from './actions';
import { filmData } from './selectors';

const mapStateToProps = createStructuredSelector({
  filmData,
});

const mapDispatchToProps = {
  setFilmId,
};

export default connect(mapStateToProps, mapDispatchToProps)(Film);
