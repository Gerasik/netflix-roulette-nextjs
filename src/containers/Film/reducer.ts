import Immutable from 'immutable';
import { REHYDRATE } from 'redux-persist/constants';

import * as Models from 'models';
import { Action, State } from './models';
import ActionType from './constants';

const initialState: State = Immutable.fromJS({});

const filmData = (state = initialState, action: Models.IAction): State => {
  switch (action.type) {
    case ActionType.SET_FILM_DATA:
      return setFilmData(state, action);

    case REHYDRATE: {
      const incoming = action.payload.filmData;
      if (incoming) return incoming;
      return state;
    }

    default:
      return state;
  }
};

export default filmData;

export const setFilmData: Models.Reducer<State, Action.SetFilmData> = (state, action) => {
  return Immutable.fromJS(action.payload);
};
