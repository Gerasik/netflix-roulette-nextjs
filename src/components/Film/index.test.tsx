import React from 'react';
import { mount } from 'enzyme';
import { createBrowserHistory } from 'history';
import { Router } from 'react-router-dom';

import { testMovieMap1, testMovieMap3 } from 'mock';
import Film from '.';

const history = createBrowserHistory();

describe('<Film />', () => {
  it('renders children when passed in', () => {
    const component = (
      <Router history={history}>
        <Film filmData={testMovieMap1} setFilmId={id => ({ type: 'string', payload: id })}>
          <div className="unique" />
        </Film>
      </Router>
    );
    const wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  });

  it('expected image placeholder and 0 runtime ', () => {
    const component = (
      <Router history={history}>
        <Film filmData={testMovieMap3} setFilmId={id => ({ type: 'string', payload: id })}>
          <div className="unique" />
        </Film>
      </Router>
    );
    const wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  });
});
