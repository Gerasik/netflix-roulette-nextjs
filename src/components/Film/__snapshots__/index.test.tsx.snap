// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Film /> expected image placeholder and 0 runtime  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Film
    filmData={
      Immutable.Map {
        "budget": 54545,
        "release_date": "2020-05-01",
        "vote_average": 55,
        "tagline": "test",
        "genres": Immutable.List [
          "drama",
        ],
        "poster_path": "",
        "title": "second movie",
        "runtime": NaN,
        "overview": "ove",
        "vote_count": 66,
        "id": 2,
        "revenue": 5454,
      }
    }
    setFilmId={[Function]}
  >
    <div
      className="container"
    >
      <picture
        className="picture"
        style={Object {}}
      >
        <source
          media="(min-width: 600px)"
          srcSet="image_placeholder.png"
        />
        <img
          alt="second movie"
          src="image_placeholder.png"
          width="50%"
        />
      </picture>
      <div
        className="info"
      >
        <div
          className="title-container"
        >
          <h1
            className="title"
          >
            second movie
          </h1>
          <div
            className="round"
          >
            <span
              className="rating"
            >
              55
            </span>
          </div>
        </div>
        <span
          className="smallDescription"
        >
          test
        </span>
        <div
          className="parameters"
        >
          <span
            className="year"
          >
            2020
          </span>
          <span
            className="runtime"
          >
            0 min
          </span>
        </div>
        <p
          className="overview"
        >
          ove
        </p>
      </div>
    </div>
  </Film>
</Router>
`;

exports[`<Film /> renders children when passed in 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Film
    filmData={
      Immutable.Map {
        "budget": 54545,
        "release_date": "2020-05-01",
        "vote_average": 55,
        "tagline": "test",
        "genres": Immutable.List [
          "drama",
          "sitcom",
          "drama",
        ],
        "poster_path": "/movie_url/image.jpg",
        "title": "first movie",
        "runtime": 5454,
        "overview": "ove",
        "vote_count": 66,
        "id": 1,
        "revenue": 5454,
      }
    }
    setFilmId={[Function]}
  >
    <div
      className="container"
    >
      <picture
        className="picture"
        style={
          Object {
            "backgroundColor": "inherit",
          }
        }
      >
        <source
          media="(min-width: 600px)"
          srcSet="/movie_url/image.jpg"
        />
        <img
          alt="first movie"
          src="/movie_url/image.jpg"
          width="100%"
        />
      </picture>
      <div
        className="info"
      >
        <div
          className="title-container"
        >
          <h1
            className="title"
          >
            first movie
          </h1>
          <div
            className="round"
          >
            <span
              className="rating"
            >
              55
            </span>
          </div>
        </div>
        <span
          className="smallDescription"
        >
          test
        </span>
        <div
          className="parameters"
        >
          <span
            className="year"
          >
            2020
          </span>
          <span
            className="runtime"
          >
            5454 min
          </span>
        </div>
        <p
          className="overview"
        >
          ove
        </p>
      </div>
    </div>
  </Film>
</Router>
`;
